WIP: Replacing Static Dataflow Analysis with Reactive Caches

hubertp opened the pull request:
### Pull Request Description

Initial implementation that replaces cache invalidation logic and static dataflow analysis with Reactive caches.

A lot of code is commented out, as either it needs to be re-done or is simply obsolete in the new implementation.
There will be still a lot of changes but basic functionality of loading/executing projects and executing visualizations works. 
Invalidation of only the necessary nodes works, by replacing static dataflow analysis with runtime analysis that collects dependencies between UUIDs.

Depends on #13907. Addresses #10525, renders #13219 obsolete.

### Important Notes

<!--
- Mention important elements of the design.
- Mention any notable changes to APIs.
-->

### Checklist

Please ensure that the following checklist has been satisfied before submitting the PR:

- [ ] The documentation has been updated, if necessary.
- [ ] Screenshots/screencasts have been attached, if there are any visual changes. For interactive or animated visual changes, a screencast is preferred.
- [ ] All code follows the
      [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md),
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md),
      [TypeScript](https://github.com/enso-org/enso/blob/develop/docs/style-guide/typescript.md),
      and
      [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md)
      style guides. In case you are using a language not listed above, follow the [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md) style guide.
- [ ] Unit tests have been written where possible.
- [ ] If meaningful changes were made to logic or tests affecting Enso Cloud integration in the libraries, 
      or the Snowflake database integration, a run of the [Extra Tests](https://github.com/enso-org/enso/actions/workflows/extra-nightly-tests.yml) has been scheduled.
  - If applicable, it is suggested to paste a link to a successful run of the Extra Tests.

hubertp committed:
WIP: Replacing Static Dataflow Analysis with Reactive Caches

Initial implementation that replaces cache invalidation logic and static
dataflow analysis with Reactive caches.

A lot of code is commented out, as either it needs to be re-done or is
simply obsolete in the new implementation.

Depends on #13907.

hubertp committed:
Fix some issues when pushing context

The remaining problems relate to the fact that we are caching all
expressions. That is conflict with the current `enterables` logic for
entering functions from a local call stack.
The latter will need a more involving rewrite.

JaroslavTulach commented on engine/runtime-instrument-common/src/main/java/org/enso/interpreter/instrument/Observable.java:
```diff
@@ -0,0 +1,115 @@
+import java.util.stream.Stream;
+import org.enso.interpreter.service.GuestExecutionService;
+
+public class Observable {
```
- could this generic(?) observable support be moved out of `runtime-instrument-common` into own project?
- it would be easier to analyze the observables system in isolation:
   - easier to unit test
   - easier to built a mental model of the observables framework
   - easier to document it
- ideally such a project would have no other dependencies than JVM
- thus the conceptual surface would be way smaller than of the whole `runtime-instrument-common` project

JaroslavTulach commented on engine/runtime-compiler/src/main/java/org/enso/compiler/pass/analyse/FramePointer.java:
```diff
@@ -8,8 +9,9 @@
-@Persistable(clazz = FramePointer.class, id = 1283)
-public record FramePointer(int parentLevel, int frameSlotIdx) implements FrameAnalysisMeta {
+@Persistable(clazz = FramePointer.class, id = 1288)
+public record FramePointer(int parentLevel, int frameSlotIdx, UUID externalId)
```
- I'd prefer `FramePointer` to remain simple pair of numbers
  - it gets serialized into the caches - it should be small
- What's the rationale for associating `externalId` with a frame pointer?
   - is it the definition ID or usage ID?

JaroslavTulach commented on engine/runtime/src/main/java/org/enso/interpreter/EnsoLanguage.java:
```diff
@@ -120,6 +121,10 @@ public final class EnsoLanguage extends TruffleLanguage<EnsoContext> {
   private final ContextLocal<ExecutionEnvironment[]> executionEnvironment =
       locals.createContextLocal(ctx -> new ExecutionEnvironment[1]);
+
+  private final ContextThreadLocal<RuntimeAnalysis> runtimeAnalysis =
```
### Rationale

- I have concerns about modifying `EnsoLanguage` and mimicking the `ExecutionEnvironment` change
- there is #12340 pending
   - describing the execution environment is done in a wrong way
- I'd rather avoid doing the same mistake for second time

### Purpose

- what's the purpose of `RuntimeAnalysis`?
- shall it behave like a `State`?
- or shall it behave like a `EnsoContext.Extra`?
- based on above answer, it then should either become part of `State` or part of `Extra`.

> /** Access to runtime analysis data associated with this context and current thread. */

- later during the review I found the above comment...
- ... then it should be part of `State`

JaroslavTulach commented on engine/polyglot-api/src/main/java/org/enso/polyglot/debugger/IdExecutionService.java:
```diff
@@ -52,7 +52,7 @@ public interface Callbacks {
      *     the execution and return the value as a result.
      */
-    Object findCachedResult(Info info);
+    Object findCachedResult(Info info, UUID downstreamDependency);
```
- I tried hard to untangle instrumentation changes with `RuntimeCache` & co.
- thus I'd like to deeply understand changes to this interface
- so far the changes here are surprising to me
- the only one that I envision here is `readingLocalVariable(someId)`
- that's a currently missing information from instrumentation to implement reactiveness

JaroslavTulach reviewed (commented):
## Inception Comments

I am glad reactive caches are shaping out. My biggest wish is to separate the `Observable` & co. into own project, provide clear API, documentation and test in isolation.

hubertp committed:
Introduce Caching and Non-caching observables

In order to be able to track dependencies between nodes, also in the
event when no external `UUID` is available, had to allow to have two types
of Observables. Similarly, `UUID`s that are external and internal are now
easily distinguishable thanks to `RuntimeID` wrapper.

Currently some tracking logic is inside nodes themselves. That's
obviously wrong - it should all be in the instrumentation. Follow up
change will move the logic. The current approach is mostly for
achieving a minimal viable proof that the approach will work.

hubertp committed:
Down to 2 failures in RuntimeServerTests

Mostly cosmetic changes due to the fact that we now invalidate less.

But two prominent changes are included:
- invalidation logic when Recompute requests comes
- Fixed tracking method calls/closure execution

The latter uses enter/exit logic in `RuntimeAnalysis` in
`InvokeCallableNode` which is correct but undeseriable. The information
should be gathered in instrumention instead. Explicit calls are easier
to experiment with and will be transfered to instrumentation eventually.

hubertp committed:
All RuntimeServerTest cases pass

Had to ensure that dependencies are being tracked correctly across
function call boundaries. This is problematic due to the fact that local
calls are being done with their own RuntimeCaches; any invalidation has
to ensure that it is capable of crossing that boundary.

Additionally, had to ensure that self argument is properly
tracked/invalidated or changes withing the functions would not be
translated into Truffle nodes.

The logic is still experiemental and elements of RuntimeAnalysis will
need to move to instrumentation, for performance reasons.

hubertp committed:
Stop `RejectedExecutionException` from showing up

The exception was being propagated, unintentionally it seems, and
showing up in logs.

hubertp committed:
Test illustrating issues in a larger project

Spurious method/self arg invalidation leading to large and unnecessary
re-compilations.

hubertp committed:
Skip tracking for synthetic node

hubertp committed:
Fix tracking for synthetic auto-scoping constructs

Unresolved constructors are being resolved in synthetic constructs.
Sadly, those constructs inside closures, share original IDs and
therefore create false dependencies between nodes.
As we still need the original IDs for instrumentation and expression
updates, introduced a flag that disables runtime tracking for
expressions.

hubertp committed:
Fixing issues in RuntimeVisualizationsTest

Not yet complete but already found a few corner cases not covered in the
new design.

hubertp committed:
Make sure CompletionExceptions are unwrapped

Extra diagnostics were lost when computation failed at an early
CompletionStage.

hubertp committed:
Show visualizations for Panics

Panics aren't cached but visualizations can still be run on them.
Had to add another method to Observable to make it happen.

hubertp committed:
Fix a number of RuntimeVisualizationTest cases

Have to keep track of arguments applied to the visualization and map
them to synthetic assignnment constructs in visualization code to get
the right identifiers.

hubertp committed:
Ensure IdMap updates can trigger invalidation

If a new external identifier is added within an expression that is
already cached, the latter should be invalidated.
Fixes a number of broken visualization tests.

hubertp committed:
No bug, just a broken test

hubertp committed:
Enable broken test by testing a workaround

When a text edit is performed on a code that is being called in the
visualization, a full re-evaluation is needed in order to generate
updated Truffle nodes.

Decided not to support that case, as it is very unlikely to be needed in
the foresable future. Text change will invalidate necessary observers,
as one would expect, but it will not trigger re-evaluation. Instead, one
has to explicitly make `ModifyVisualization` request.
This seems like a good compromise as evaluations of visualization
expressions are very costly (also due to locking).

hubertp committed:
Fix some regressions

Workarounds for Stackoverflow issues + IdMap updates causing full
program invalidations. This change breaks 1 test in
RuntimeVisualizationTests that will be addressed separately.

hubertp committed:
Fix a broken assumption with local variables access

With runtime tracking we are able to precisely discover which cached
values need to be invalidated. Unfortunately excessive use of external
UUIDs (and therefore caching) by GUI revealed a flaw in that logic wrt
to local variable access.

Whenever a local variable is being accessed it is reading a previously
written value from a frame. But if the assignment has been cached, then
a specific slot in the frame is not written during successive execution.
This means that adding new expressions that make use of previously
cached local variables would report uninitialized value errors.

This change workarounds this by **never** caching assignments.
Assignments should be fast if the RHS is already cached, and GUI appears
to assign external UUIDS to RHS always, it seems. That way any reading
from a frame's slot will succeed.

Also reverted to the old tree building in Changeset as it appeared to be
sufficient for our needs.

hubertp committed:
Renaming

hubertp committed:
Fix RuntimeErrorsTest

Old change disabled caching of DataflowErros. That does not appear to be
necessary anymore.
Also disovered a current limitation of the approach when an edit is
being made in a local function that is different from the current
execution context. In that case we are not able to properly
track/invalidate dependencies thus leading to stale computations.
This is hypothetical scenario as GUI would never make such a call.

hubertp committed:
Adapting RuntimeRefactoringTest

`RuntimeRefactoringTest` tracked corrected by making sure that the
function/variable symbol is being set as pending.
That feels like it was by accident by expression updates should only
list expressions and currently we would no longer list those.
Enhanced tests to make sure that change propagation is correctly
triggered.

hubertp committed:
Merge branch 'develop' into wip/hubert/10525-reactive-cache-take-2

hubertp committed:
Merge branch 'develop' into wip/hubert/10525-reactive-cache-take-2

hubertp committed:
Merge branch 'develop' into wip/hubert/10525-reactive-cache-take-2

hubertp commented:
Superseded by #14633
